packages/apps/plugins/plugin-graph/src/graph/types.ts

```diff
@@ -0,0 +1,71 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import type { IconProps } from '@phosphor-icons/react';
+import { FC } from 'react';
+
+import type { Intent } from '@braneframe/plugin-intent';
+import type { UnsubscribeCallback } from '@dxos/async';
+
+export interface Graph {
+  get root(): Graph.Node;
+  find(id: string): Graph.Node | undefined;
+  traverse(options: Graph.TraverseOptions): void;
+  registerNodeBuilder(builder: Graph.NodeBuilder): void;
+  removeNodeBuilder(builder: Graph.NodeBuilder): void;
+  construct(): void;
+}
+
+export namespace Graph {
+  export type NodeBuilder = (parent: Node) => UnsubscribeCallback | void;
+
+  export type TraverseOptions = {
+    from?: Node;
+    direction?: 'up' | 'down';
+    predicate?: (node: Node) => boolean;
+    onVisitNode?: (node: Node) => void;
+  };
+
+  export type Node<TData = any, TProperties extends { [key: string]: any } = { [key: string]: any }> = {
```

Comment:
for what purpose would we want the types to be separate?

Reply:
Rich and I discussed this, It boiled down to having the ability to write tight utilities against the GraphNode interface without "bringing along all of it". for example, utilities can be like `iCanReorderAThing(node: IOrderable) {}` where `GraphNode = IOrderable & IHaveChildren & ...`. I think two things should obviate the need to do anything about this now:
1. we haven't actually felt any pain from using just plain `GraphNode` interface in all such cases (i.e. let's just write such utilities against the full-sized GraphNode and see if that hurts anywhere to perform this surgery.
2. any such utility can just declare locally the signature subset of GraphNode that it cares about (in case we don't get around to redeclaring GraphNode as an algebraic sum before we need this). I.e. `INeedOrderable(node: { properties: { index: number })` and regular `GraphNodes` will be type-compatible with this either way.

It also helps to clarify any concrete scenario we have in front of us to motivate changes like this.

Reply:
yeah I can see how that would be useful given a motivating use case :+1: